fix: include team_id in Person queries for partitioned table compatibility

webjunkie opened the pull request:
## Problem

The new `posthog_person_new` table is hash-partitioned by `team_id` with 64 partitions:
- Primary key: `(team_id, id)`
- UUID index: `(team_id, uuid)`

Queries without `team_id` must scan all 64 partitions instead of 1, causing ~64x performance degradation.

## Changes

**1. Fix split_person Celery task** (posthog/tasks/split_person.py)
- Add `team_id` parameter to task signature
- Change query from `Person.objects.get(pk=person_id)` to `Person.objects.get(team_id=team_id, pk=person_id)`

**2. Fix Person.split_person() method** (posthog/models/person/person.py)
- Change `Person.objects.get(pk=self.pk)` to `Person.objects.get(team_id=self.team_id, pk=self.pk)`
- `self.team_id` is always available on person instances

**3. Update API endpoint caller** (posthog/api/person.py)
- Pass `person.team_id` as second parameter to split_person task
- Change from `split_person.delay(person.id, ...)` to `split_person.delay(person.id, person.team_id, ...)`

**4. Fix batch deletion** (posthog/models/team/util.py)
- Preserve `team_id` filter in `_raw_delete_batch()` when creating batch queryset
- Prevents scanning all partitions during team deletion

## Impact

- Split person operations: 1 partition scan instead of 64 (~64x faster)
- Team deletion: Only relevant partition(s) scanned instead of all 64
- No breaking changes - all modified code already has access to team_id

## Testing

**All other production Person queries already optimized:**
- ✅ PersonViewSet auto-filters by team_id via `TeamAndOrgViewSetMixin`
- ✅ Session recordings API already includes `team=team` filter
- ✅ All raw SQL queries already include `team_id = %(team_id)s`
- ✅ Signal handlers have access to `instance.team.id`

**Verification done:**
- Audited all Person queries in posthog/api/, posthog/models/, posthog/tasks/
- Audited session recordings, HogQL, feature flags, temporal workflows, ee/
- Only these 3 production code paths needed fixes

## Context

This prepares the codebase for cutover from `posthog_person` to the partitioned `posthog_person_new` table. The partitioning strategy uses hash partitioning on `team_id` to distribute data across 64 partitions for better scalability.

webjunkie committed:
fix: include team_id in Person queries for partitioned table compatibility

Prepare Person model queries for hash-partitioned posthog_person_new table (64 partitions by team_id).

Without team_id in queries, PostgreSQL must scan all 64 partitions instead of 1, causing ~64x performance degradation.

**Changes:**

- Add team_id parameter to split_person Celery task
- Include team_id in Person.split_person() internal query
- Update API endpoint to pass team_id when calling split_person task
- Fix _raw_delete_batch() to preserve team_id filter during batch deletions

**Impact:**

- Split person operations will scan 1 partition instead of 64
- Team deletion will scan only relevant partitions instead of all 64
- No breaking changes - all modified code paths already have team_id available

**Context:**

The new posthog_person_new table uses hash partitioning:
- Primary key: (team_id, id)
- UUID index: (team_id, uuid)
- 64 hash partitions by team_id

Most queries already include team_id via ViewSet filtering or ForeignKey relationships. These fixes address the remaining production code paths that used PK-only lookups.

benjackwhite commented on posthog/tasks/split_person.py:
Although you need to modify the caller in the api code (search `split_person.delay`)

webjunkie replied:
That is done in posthog/api/person.py as far as I can see.

webjunkie commented on posthog/models/person/person.py:
```diff
@@ -86,7 +86,7 @@ def _add_distinct_ids(self, distinct_ids: list[str]) -> None:
             self.add_distinct_id(distinct_id)
 
     def split_person(self, main_distinct_id: Optional[str], max_splits: Optional[int] = None):
```
I think from the API we have the base classes adding the team_id (for safety reasons already), but I can double check.

webjunkie committed:
fix: address review feedback for partitioned table queries

**1. Backward compatibility for split_person task**

Add support for old 3-arg signature during rolling deploys.
Detect legacy calls by checking if team_id arg is a string (old main_distinct_id position).
Falls back to non-partitioned query for old tasks, uses optimized query for new ones.

**2. Fix API delete endpoint to include team_id**

Override perform_destroy() to use queryset delete instead of instance delete.
Django's instance.delete() generates: DELETE FROM posthog_person WHERE id = X
This scans all 64 partitions without team_id filter.

New queryset delete: Person.objects.filter(team_id=Y, id=X).delete()
Scans only 1 partition instead of 64.

Fixes both destroy() and bulk_delete() endpoints.

webjunkie committed:
fix: mypy type errors in split_person backward compatibility

github-actions[bot] committed:
Update query snapshots

eli-r-ph committed:
fix(django): add team_id to only clause in actor_base_query.py (#41654)

Co-authored-by: Robbie Coomber <robbie.coomber@gmail.com>

webjunkie committed:
feat: add configurable Person table via PERSON_TABLE_NAME env var

Add ability to switch Person model between posthog_person (legacy) and
posthog_person_new (partitioned) tables via environment variable.

**Changes:**
- Add PERSON_TABLE_NAME setting in data_stores.py (defaults to posthog_person)
- Set Person.Meta.db_table = settings.PERSON_TABLE_NAME for dynamic table name
- Add state-only migration to track Meta change in Django migration system
- Add smoke test to verify Person model uses configured table name

**Usage:**
- Default: uses posthog_person (no env var needed)
- For partitioned table: set PERSON_TABLE_NAME=posthog_person_new
- Table name is evaluated at Django startup, no runtime overhead

Stacked on: fix/person-queries-include-team-id-for-partitioning

webjunkie committed:
feat: add configurable Person table via PERSON_TABLE_NAME env var

Add ability to switch Person model between posthog_person (legacy) and
posthog_person_new (partitioned) tables via environment variable.

**Changes:**
- Add PERSON_TABLE_NAME setting in data_stores.py (defaults to posthog_person)
- Set Person.Meta.db_table = settings.PERSON_TABLE_NAME for dynamic table name
- Add state-only migration to track Meta change in Django migration system
- Add smoke test to verify Person model uses configured table name

**Usage:**
- Default: uses posthog_person (no env var needed)
- For partitioned table: set PERSON_TABLE_NAME=posthog_person_new
- Table name is evaluated at Django startup, no runtime overhead

Stacked on: fix/person-queries-include-team-id-for-partitioning

eli-r-ph committed:
fix(django): ensure person DELETE is scoped by team_id in _raw_delete_batch (#41660)

webjunkie committed:
feat: add fail-loud team_id enforcement to Person QuerySet

Add PersonQuerySet that intercepts all query execution and validates that
team_id is present in the WHERE clause. This is critical for the partitioned
posthog_person_new table (64 hash partitions by team_id) - queries without
team_id would scan all 64 partitions causing ~64x performance degradation.

**Implementation:**
- PersonQuerySet._fetch_all() intercepts before any query execution
- _has_team_id_filter() checks if team_id appears in SQL WHERE clause
- Raises ValueError with clear message and SQL snippet if team_id missing
- PersonManager.get_queryset() returns PersonQuerySet for all Person queries

**Enforcement catches ALL QuerySet operations:**
- get(), filter(), exclude(), update(), delete()
- count(), exists(), aggregate()
- Iteration, slicing, indexing
- Any chained QuerySet method

**Benefits:**
- Permanent safety mechanism (no opt-out needed)
- Catches bugs in tests AND production code paths
- Clear error messages for debugging
- Minimal performance overhead (single check per query evaluation)

**Changes:**
- Add PersonQuerySet class with team_id validation
- Update PersonManager to use PersonQuerySet
- Fix test bug: Person.objects.all() → Person.objects.filter(team_id=...)
- Add smoke test verifying enforcement raises ValueError

Stacked on: fix/person-queries-include-team-id-for-partitioning

webjunkie committed:
feat: add fail-loud team_id enforcement to Person QuerySet

Add PersonQuerySet that intercepts all query execution and validates that
team_id is present in the WHERE clause. This is critical for the partitioned
posthog_person_new table (64 hash partitions by team_id) - queries without
team_id would scan all 64 partitions causing ~64x performance degradation.

**Implementation:**
- PersonQuerySet._fetch_all() intercepts before any query execution
- _has_team_id_filter() checks if team_id appears in SQL WHERE clause
- Raises ValueError with clear message and SQL snippet if team_id missing
- PersonManager.get_queryset() returns PersonQuerySet for all Person queries

**Enforcement catches ALL QuerySet operations:**
- get(), filter(), exclude(), update(), delete()
- count(), exists(), aggregate()
- Iteration, slicing, indexing
- Any chained QuerySet method

**Benefits:**
- Permanent safety mechanism (no opt-out needed)
- Catches bugs in tests AND production code paths
- Clear error messages for debugging
- Minimal performance overhead (single check per query evaluation)

**Changes:**
- Add PersonQuerySet class with team_id validation
- Update PersonManager to use PersonQuerySet
- Fix test bug: Person.objects.all() → Person.objects.filter(team_id=...)
- Add smoke test verifying enforcement raises ValueError

Stacked on: fix/person-queries-include-team-id-for-partitioning

webjunkie committed:
fix: handle EmptyResultSet exception in team_id filter check

Django raises EmptyResultSet when a query's WHERE clause evaluates to always-false
(e.g., WHERE 0=1). These queries are safe - they return no results without scanning
partitions. Allow them through by catching EmptyResultSet and returning True.

Fixes test failures in test_action_no_steps and test_filter_by_nonexisting_person.

webjunkie committed:
fix: handle EmptyResultSet exception in team_id filter check

Django raises EmptyResultSet when a query's WHERE clause evaluates to always-false
(e.g., WHERE 0=1). These queries are safe - they return no results without scanning
partitions. Allow them through by catching EmptyResultSet and returning True.

Fixes test failures in test_action_no_steps and test_filter_by_nonexisting_person.

webjunkie committed:
fix: add team_id filters to Person queries in test files

Add team_id filters to all Person model queries across test files to ensure
compatibility with partitioned posthog_person_new table (64 hash partitions by team_id).

Changes:
- Session recording tests: Replace Person.objects.all().delete() with
  Person.objects.filter(team_id__isnull=False).delete() in setUp methods
- Cohort tests: Add team_id=cohort.team_id to Person.objects.filter(cohort__id=...) queries
- Person API tests: Add team_id filters to Person.objects.all().order_by() queries
- Events query runner test: Add team_id filter to Person.objects.first()

Without team_id in WHERE clause, these queries would scan all 64 partitions
causing ~64x performance degradation. While these are test-only code paths,
the patterns represent real bugs that would occur in production if similar
code existed there.

Related to partitioned table enforcement implemented in PersonQuerySet._fetch_all().

benjackwhite committed:
Override of delete function

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' of github.com:PostHog/posthog into fix/person-queries-include-team-id-for-partitioning

benjackwhite committed:
Fix

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' into feat/configurable-person-table

# Conflicts:
#	posthog/models/person/person.py

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' into feat/person-queryset-team-id-enforcement

# Conflicts:
#	posthog/models/person/person.py

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Fixes

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' of github.com:PostHog/posthog into fix/person-queries-include-team-id-for-partitioning

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' into feat/configurable-person-table

webjunkie committed:
feat: add delete() and update() enforcement to PersonQuerySet

Extend team_id enforcement to DELETE and UPDATE operations on Person model.
Previously only SELECT queries were validated via _fetch_all().

Changes:
- Override delete() to validate team_id filter before executing DELETE
- Override update() to validate team_id filter before executing UPDATE
- Add test_person_delete_enforces_team_id test
- Add test_person_update_enforces_team_id test

Without team_id in WHERE clause, DELETE and UPDATE operations would scan
all 64 partitions of the partitioned posthog_person_new table.

The delete() enforcement caught real bugs in test setUp methods that were
already fixed in previous commit (Person.objects.all().delete() changed to
Person.objects.filter(team_id__isnull=False).delete()).

benjackwhite committed:
FixFix

webjunkie committed:
fix: replace hardcoded table names with _meta.db_table in SQL queries

Replace hardcoded 'posthog_person' and 'posthog_persondistinctid' with dynamic
table names from model _meta.db_table. This ensures queries work correctly when
table names change via PERSON_TABLE_NAME configuration.

Files updated:
- posthog/models/cohort/cohort.py: Person table in SQL string replacement
- posthog/hogql_queries/actor_strategies.py: Person and PersonDistinctId queries
- posthog/models/feature_flag/flag_matching.py: PersonDistinctId and Person queries

All files already had the models imported or added them for table name access.

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' into feat/configurable-person-table

# Conflicts:
#	posthog/models/person/person.py

benjackwhite committed:
Merge branch 'feat/person-queryset-team-id-enforcement' of github.com:PostHog/posthog into feat/person-queryset-team-id-enforcement

# Conflicts:
#	posthog/models/person/person.py
#	posthog/models/test/test_person_schema.py

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' into feat/person-queryset-team-id-enforcement

# Conflicts:
#	posthog/models/person/person.py

webjunkie committed:
fix: parameterize table names in management commands and temporal workflows

Replace hardcoded 'posthog_person' with Person._meta.db_table in management
commands and temporal workflows. These were previously parameterized but
reverted in commit d3916cb4a6 when switching back to posthog_person as
standard name.

Re-parameterizing now to support dynamic table names via configuration.

Files updated:
- posthog/management/commands/delete_persons.py
- posthog/management/commands/delete_persons_with_no_distinct_ids.py
- posthog/management/commands/fix_cohort_people_unsynced_ch.py
- posthog/temporal/delete_persons/delete_persons_workflow.py

All files now import Person model and use _meta.db_table for SQL queries.

webjunkie committed:
Merge branch 'feat/person-queryset-team-id-enforcement' into fix/person-queries-include-team-id-for-partitioning

webjunkie committed:
fix: prevent Person queries without team_id from lazy FK resolution

Add select_related('person') to PersonDistinctId queries to prevent
lazy Foreign Key loading that generates queries without team_id filter.

Changes:
- sync_persons_to_clickhouse: Add select_related to avoid lazy loading
- deletion.py: Same fix for person reset operations
- session_recording_api: Replace select_related with prefetch_related
  using explicit Prefetch with team_id filter to ensure JOIN includes
  team_id in partition pruning
- split_person task: Add backward compatibility lookup via PersonDistinctId
  to get team_id before querying Person table
- split_person command: Add team_id to Person.objects.get() call

Without these fixes, accessing person_distinct_id.person.uuid triggers:
SELECT * FROM posthog_person_new WHERE id = X LIMIT 21
which scans all 64 partitions instead of single partition.

webjunkie committed:
fix: improve team_id enforcement to check WHERE clause only

Previous implementation checked if "team_id" appeared anywhere in the
SQL string, which incorrectly passed queries that had team_id in the
SELECT clause but not in the WHERE clause.

For example, this query would have passed the check:
  SELECT id, team_id FROM posthog_person_new WHERE id = 123 LIMIT 21

Even though it scans all 64 partitions because the WHERE clause lacks
team_id filter for partition pruning.

Now extracts the WHERE clause portion and checks specifically for
team_id presence there, correctly catching lazy FK resolution queries.

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
fix

benjackwhite committed:
Merge branch 'fix/person-queries-include-team-id-for-partitioning' into feat/configurable-person-table

# Conflicts:
#	posthog/models/person/person.py
#	posthog/models/test/test_person_schema.py

benjackwhite committed:
fix

eli-r-ph committed:
fix delete query to utilize (team_id, id) index on persons

github-actions[bot] committed:
Update query snapshots
